client/src/route/auth/Functions.js

```diff
@@ -0,0 +1,194 @@
+
+    return [accessToken, newRefreshToken, expiresIn];
+  } catch (error) {
+    console.error(`Error refreshing access token: ${error}`);
+    throw error;
+  }
+}
+
+/**
+ * Requests a GitLab OAuth access token using an authorization code.
+ *
+ * @param {string} clientId - The client ID for the GitLab OAuth application.
+ * @param {string} clientSecret - The client secret for the GitLab OAuth application.
+ * @param {string} redirectUri - The redirect URI for the GitLab OAuth application.
+ * @param {string} authCode - The authorization code received from GitLab after the user grants permission.
+ * @throws {Error} If the token request fails.
+ * @returns {Promise<[string, string] | null>} - A Promise that resolves to an array containing the access token and refresh token, or null if the request failed.
+ */
+export async function getGitLabAccessToken(
+  clientId,
+  clientSecret,
+  redirectUri,
+  authCode,
+) {
+  try {
+    // Request access token using authorization code
+    const response = await fetch('https://gitlab.com/oauth/token', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: new URLSearchParams({
+        grant_type: 'authorization_code',
+        client_id: clientId,
+        client_secret: clientSecret,
+        redirect_uri: redirectUri,
+        code: authCode,
+      }),
+    });
+
+    if (!response.ok) {
+      throw new Error(
+        `Error fetching access token: ${response.statusText}`,
+      );
+    }
+
+    const data = await response.json();
+    const accessToken = data.access_token;
+    const refreshToken = data.refresh_token;
+    const expiresIn = data.expires_in;
+
+    return [accessToken, refreshToken, expiresIn];
+  } catch (error) {
+    console.error(error);
+    return null;
+  }
+}
+
+/**
+A function that fetches data from the GitLab GraphQL API using an access token and a GraphQL query.
+@param {string} accessToken - The GitLab access token to use for the API request.
+@param {string} query - The GraphQL query to execute on the GitLab API.
+@returns {object} The response data from the GraphQL query as an object.
+@throws {Error} If an error occurs while fetching data from the GitLab GraphQL API.
+@returns {null} If the API request fails.
+*/
+export async function fetchGitLabGraphQLQuery(accessToken, query) {
+  try {
+    const response = await fetch('https://gitlab.com/api/graphql', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        // eslint-disable-next-line quote-props
+        Authorization: `Bearer ${accessToken}`,
+      },
+      body: JSON.stringify({ query }),
+    });
+    if (!response.ok) {
+      throw new Error(
+        `Error fetching GitLab GraphQL data: ${response.statusText}`,
+      );
+    }
+    const data = await response.json();
+    return data;
+  } catch (error) {
+    console.error(error);
+    return null;
+  }
+}
```

Comment:
- Please convert to typescript
- Take the gitlab url from the react env file
- Either code first or schema first, but please specify and bring out the graphQL schema from the code